File: 31-60/nQueens51.py

def solveNQueens(n) -> list:
    col = set()
    posDiag = set()
    negDiag = set()

    res = []
    board = [["."] * n for _ in range(n)]
    print('board', board)

    def backtrack(r):
        print('r from the start of the backtrack', r)
        if r == n:
            copy = ["".join(row) for row in board]
            print('copy', copy)
            res.append(copy)
            return
        
        for c in range(n):
            print('c', c)
            print('ccccccccccccc', c, 'col', col,'r', r, 'r+c', r+c, 'posDiag', posDiag, 'r', r, 'r-c', r-c, 'negDiag', negDiag)
            print('true or false', c in col or (r+c) in posDiag or (r-c) in negDiag)
            if c in col or (r+c) in posDiag or (r-c) in negDiag:
                continue

            col.add(c)
            posDiag.add(r+c)
            negDiag.add(r-c)
            board[r][c] = "Q"
            print('colllllllllllllllllll', col, 'r', r, 'c', c, '[r][c]', board[r][c], 'board', board, 'posDiag', posDiag, 'negDiag', negDiag, 'Start backtrack...')
            backtrack( r + 1)

            col.remove(c)
            posDiag.remove(r+c)
            negDiag.remove(r-c)
            board[r][c] = "."
            print('rrrrrrrrrrrrrrafter remove:', board)
            
    backtrack(0)
    return res



print(solveNQueens(4))

File: 31-60/nQueens51.js
function nQueens(n) {
    let col = new Set();
    let posDiag = new Set();
    let negDiag = new Set();

    let board = Array.from({length: n}, () => Array(n).fill('.'));
    let res = [];

    function backtracking(r) {

        console.log('starting a new row', r)
        if (r == n) {
           res = board.map(row => row.join(''));
           console.log('res', res)
        }
        

        for (let c = 0; c < n; c++) {
            console.log('True or False: ', col.has(c) || posDiag.has(r+c)||negDiag.has(r-c), 'c',c, 'col', col, 'r+c', r+c, 'posDiag', posDiag, 'r-c', r-c, 'negDiag', negDiag);
            if(col.has(c) || posDiag.has(r+c)||negDiag.has(r-c)) continue;

            col.add(c);
            posDiag.add(r+c);
            negDiag.add(r-c);
            board[r][c] = 'Q';

            console.log('row', r, 'column', c, 'posDiag', posDiag, 'negDiag', negDiag, 'board', board);

            backtracking(r+1);

            console.log('backing to the previous row', r);
            col.delete(c);
            posDiag.delete(r+c);
            negDiag.delete(r-c);
            board[r][c] = '.';
            console.log('row111', r, 'column111', c, 'posDiag111', posDiag, 'negDiag111', negDiag, 'board111', board);

        }
    }

    backtracking(0);
    return res;

}

console.log('nQueens', nQueens(4));